baseapp/baseapp.go

```diff
@@ -623,16 +646,32 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		return err.Result()
 	}
 
-	// run the ante handler
+	// Execute the ante handler if one is defined. Note that during deliverTx, if
+	// execution fails and thus needs to abort, state modifications will not be
+	// persisted due to the transient CacheWrap multi-store that is used. Otherwise,
+	// during checkTx, state modifications are persisted between respective checkTx
+	// calls.
 	if app.anteHandler != nil {
-		newCtx, result, abort := app.anteHandler(ctx, tx, (mode == runTxModeSimulate))
+		var msCache sdk.CacheMultiStore
+
+		anteCtx := ctx
+		if mode == runTxModeDeliver {
```

Comment:
Why don't we need to cache for runTxModeCheck?

Reply:
Perhaps we do - if we do that, we also need to `.write()` on `runTxModeCheck` - this PR just didn't change that because the previous behavior was to apply `CheckTx()` changes in the ante handler without cache-wrapping.